Cfg: Add support to reset admin password at startup (setting: `reset_admin_password_at_start`)

BobDu opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #49055
I can not reopen #55980 . Reopen a new PR.

**Special notes for your reviewer:**

Please check that:
- [x] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

BobDu committed:
add reset_admin_every_time_run in cfg security

Signed-off-by: BobDu <i@bobdu.cc>

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -602,7 +602,17 @@ Default is `admin`.
+
+### reset_admin_every_time_run
+
+> Only available in Grafana v10.1+
```
```suggestion
> Available in Grafana v10.1+
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -602,7 +602,17 @@ Default is `admin`.
+
+> Only available in Grafana v10.1+
+
+Reset the admin user and admin password every time start grafana server. Default is `false`.
```
```suggestion
Resets the Grafana Admin user and password every time the Grafana server starts. The default value is `false`.
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -602,7 +602,17 @@ Default is `admin`.
+Default is `admin`.
+
+> In Grafana < v10.1, only set once on first-run.
+> In Grafana v10.1+, behavior depends on `reset_admin_every_time_run`. When `reset_admin_every_time_run`=`false`, only set once on first-run.
```
```suggestion
> In Grafana v10.1+, the behavior depends on the `reset_admin_every_time_run` value. When `reset_admin_every_time_run`is `false`, the password is set once on first run.
```

chri2547 reviewed (commented):
Added some suggestions to the docs update. Please review before you accept to ensure I didn't change the technical meaning of what you wrote. I'll let engineering weigh in on the code updates. Thank you very much for your contribution!

BobDu committed:
add reset_admin_every_time_run in cfg security

Signed-off-by: BobDu <i@bobdu.cc>

BobDu commented:
@chri2547  Tks. I have modified as you suggested.

BobDu commented:
Do you have any more suggestions for cfg option names?
Is there a better fit than `reset_admin_every_time_run `?
@chri2547

chri2547 commented:
> Do you have any more suggestions for cfg option names? Is there a better fit than `reset_admin_every_time_run `? @chri2547

Good question! I'll let engineering address your question. They might have a particular naming convention they'd like to use.

mildwonkey commented:
@BobDu if you're looking for nitpicking, I would go with  `reset_admin_password_at_start` or even just `reset_admin_password` and note that the admin password will reset every time grafana starts up if that's true.

BobDu committed:
add reset_admin_password_at_start in cfg security

Signed-off-by: BobDu <i@bobdu.cc>

BobDu commented:
modified use cfg opt name `reset_admin_password_at_start `

mildwonkey reviewed (dismissed):
Thank you so much for getting back to this, @BobDu !  

This isn't quite working correctly - I'm writing up my review now, but wanted to get this in place first :)

mildwonkey commented on pkg/services/sqlstore/user.go:
```diff
@@ -132,6 +132,37 @@ func (ss *SQLStore) createUser(ctx context.Context, sess *DBSession, args user.C
+		usr.Password = encodedPassword
+	}
+
+	_, err = sess.ID(1).Update(usr)
```
It's not safe to assume that the admin user ID is always 1 - it most likely is, especially in the use cases you had in mind when adding this configuration, but it could be used in an instance where the admin user ID is not 1. We will need to verify that user is the admin. 

You can compare this to the reset-admin-password CLI command, which also assumes `1` is the default admin user ID, but validates that by first getting the user and checking the is_admin field: https://github.com/grafana/grafana/blob/main/pkg/cmd/grafana-cli/commands/reset_password_command.go#L53

Also, take a look at the user service [UpdatePassword](https://github.com/grafana/grafana/blob/main/pkg/services/user/userimpl/store.go#L366) method; your Update command should look closer to that one.

----

Please let me know if you need more specific suggestions, or even if you want me to push a commit to your PR branch - I don't want to over-explain and I'm not in a rush, but I also don't want you to feel abandoned (again) :)

BobDu replied:
It would be really nice if you could push commit to the branch.

I just hope we can have this feature soon

mildwonkey replied:
Sure thing! Just make sure you've [Allowed edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) on your fork and I'll see what I can do 😁

mildwonkey committed:
minor fixes:
- return after resetting admin password
- verify that user id 1 is an admin user before resetting.

mildwonkey commented:
Note: the "Check for patch conflicts" failure is expected - it can't run off a fork - and is not an issue.

mildwonkey commented:
(don't mind me repeatedly changing the issue title - it's relevant for the changelog, and this hard to categorize)
